flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SystemProcessingTimeServiceTest.java

```diff
@@ -154,27 +152,25 @@ public void testImmediateShutdown() throws Exception {
 			// the task should trigger immediately and should block until terminated with interruption
 			timer.registerTimer(System.currentTimeMillis(), timestamp -> {
 				synchronized (lock) {
-					latch.trigger();
-					Thread.sleep(100000000);
+					try {
+						latch.trigger();
+						Thread.sleep(100000000);
+						fail("should be interrupted");
+					} catch (InterruptedException ignored) {
+					}
 				}
 			});
 
 			latch.await();
 			timer.shutdownService();
+
+			// can only enter this scope after the task is interrupted
 			synchronized (lock) {
 				assertTrue(timer.isTerminated());
-
-				// The shutdownService() may not necessary wait for active tasks to finish properly.
-				// From the ScheduledThreadPoolExecutor Java docs:
-				// 	There are no guarantees beyond best-effort attempts to stop processing actively executing tasks.
-				// 	This implementation cancels tasks via {@link Thread#interrupt}, so any task that
-				// 	fails to respond to interrupts may never terminate.
-				assertThat(errorRef.get(), is(anyOf(nullValue(), instanceOf(InterruptedException.class))));
```

Comment:
I don't understand this change. Shouldn't the `InterruptedException` not be a problem here because we allow it to be contained in `errorRef` here? Moreover, according to the JIRA description the problem with the `InterruptedException` failing the assertion occurs in line 196.

Reply:
Well, the problem is related to `synchronized` blocks: as soon as the sleep is interrupted, the monitor in the `synchronized` block in the task (line 154) is released **immediately**  allowing the main thread enter it's `synchronized` block (line 168) not waiting for exception being handled inside the pool thread. So sometimes when the main thread reaches assertion on line 172, the pool thread is still not finished with putting `InterruptedException` inside `errorRef`. Thus `is(anyOf...` assertion is needed.
 
The main problem is that sometimes (very rarely) the exception handling inside the pool thread is so slow, that `errorRef` gets updated with `InterruptedException` **after** line 176 (`errorRef.set(null);`) and thus making the assertion on line 196 fail. 

My changes ensure that the exception will be handled inside the task in the `synchronized` block thus releasing the monitor only after exception handling is finished. In this case the `synchronized` blocks will work as expected (as I see it) and the two threads will be really synchronized. This means that `errorRef` will always contain null. That's why I removed  `is(anyOf...` assertion here.

Another possible solution is to leave everything as it was and only change the following assertions (on line 196 and the next one) to `is(anyOf...` type. But in this case I don't really understand what the intention of 'synchronized` blocks is - because they doesn't really help with thread synchronization...

Reply:
Thanks for the clarification @a-siuniaev. Now the change makes sense to me.

Reply:
@a-siuniaev, thanks for explanation and the fix.

Just for information, the test was checking that the first timer callback may report exception (via callback, although with wrong synchronization). But the second and third timer callbacks won't use exception reporting mechanism, but rather throw at timer registration call. This is basically the reason why the assert that checks error was pulled up, near the first timer and for other timers, the similar check was verifying that there is no reported error.

In some tests, the error was `null`  and my assumption was that thread could have "interrupted" by force (without allowing to call any follow up code. But looks like it was rather a data race).
An alternative fix could be to move setting `errorRef` under the same critical section.